src/kpro_connection.erl

```diff
@@ -161,6 +163,30 @@ stop(Pid) when is_pid(Pid) ->
 stop(_) ->
   ok.
 
+%% @doc Reauthenticates SASL by repeating the authentication flow after given time.
+%% The authentication flow is repeated only once and not periodically.
+%% This would be called by an authentication adapter to reauthenticate before
+%% session_lifetime_ms provided in the v1 SASL authentication response is
+%% reached.
+%% Example use case:
+%%   -module(my_custom_sasl_authentication).
+%%
+%%   auth(Host, Sock, Vsn, Mod, ClientId, Timeout, Opts) ->
+%%     {ok, SaslResponse} = do_authenticate(...),
+%%     case session_lifetime_ms(SaslResponse) of
+%%       SessionLifetime when SessionLifetime > 0 ->
+%%         kpro_connection:sasl_reauthenticate_after(self(), a_bit_before(SessionLifetime));
+%%       _ ->
+%%         ok
+%%     end,
+%%     ok.
+-spec sasl_reauthenticate_after(connection(), timeout()) -> ok.
+sasl_reauthenticate_after(Pid, Time) when is_pid(Pid) ->
+  erlang:send_after(Time, Pid, sasl_reauthenticate),
+  ok;
+sasl_reauthenticate_after(_, _) ->
+  ok.
+
```

Comment:
It will depend on the taste of the maintainers of `kafka_protocol` ultimately, but this could also just be `sasl_reauthenticate` and leave the timer-ing to the caller.

Reply:
Yes, precisely because it needs to send a precise message, as the message atom is part of the private API.
So it would need to be `erlang:send_after(Time, Connection, kpro_connection:make_sasl_reauthenticate_msg())`